### PR TITLE
Drop support for old Rails versions < 4.2

### PR DESCRIPTION
**Pull request by bquorning**
Can active_record_shards drop support for Rails 3.2 yet?

### CODE REVIEW

**Comment by grosser** [comment]
https://github.com/zendesk/active_record_shards/tree/v3.11.5 should work

**Comment by mstroming** [comment]
Is it also the case that this gem only works with Ruby 2.2 and up?

**Comment by grosser** [comment]
yeah most likely ... nobody should be running ruby <2.2 anyway ... it's dead :)

File: lib/active_record_shards.rb [hunk 1]
```diff
@@ -24,10 +25,10 @@ def self.rails_env
 ActiveRecord::Base.extend(ActiveRecordShards::DefaultSlavePatches)
 ActiveRecord::Relation.include(ActiveRecordShards::DefaultSlavePatches::ActiveRelationPatches)
 ActiveRecord::Associations::CollectionProxy.include(ActiveRecordShards::AssociationCollectionConnectionSelection)
+ActiveRecord::Associations::Builder::HasAndBelongsToMany.include(ActiveRecordShards::DefaultSlavePatches::Rails41HasAndBelongsToManyBuilderExtension)
+ActiveRecord::SchemaDumper.prepend(ActiveRecordShards::SchemaDumperExtension)
```
Comment (bquorning) [inline lib/active_record_shards.rb]: These lines are moved from lib/active_record_shards/patches-4-2.rb and lib/active_record_shards/patches-5-0.rb

**Comment by bquorning** [comment]
@pschambacher @grosser Another review, please? I’ve tried to clean up as much as possible this time around.

File: README.md [hunk 2]
```diff
@@ -5,7 +5,7 @@
 ActiveRecord Shards is an extension for ActiveRecord that provides support for sharded database and slaves. Basically it is just a nice way to
 switch between database connections. We've made the implementation very small, and have tried not to reinvent any wheels already present in ActiveRecord.
 
-ActiveRecord Shards has been used and tested on Rails 3.2, 4.2 and 5.0 and has in some form or another been used in production on a large Rails app for several years.
+ActiveRecord Shards has been used and tested on Rails 4.2 and 5.0 and has in some form or another been used in production on a large Rails app for several years.
```
Comment (pschambacher) [inline README.md]: You _could_ add 5.1.

**Comment by craig-day** [comment]
@bquorning can we merge this?

File: lib/active_record_shards/connection_switcher-4-0.rb [hunk 3]
```diff
@@ -58,11 +47,7 @@ def connection_pool_key
...
 
     def connected_to_shard?
-      if ActiveRecord::VERSION::MAJOR >= 4
-        specs_to_pools = Hash[connection_handler.connection_pool_list.map { |pool| [pool.spec, pool] }]
-      else
-        specs_to_pools = connection_handler.connection_pools
-      end
+      specs_to_pools = Hash[connection_handler.connection_pool_list.map { |pool| [pool.spec, pool] }]
 
       specs_to_pools.key?(connection_pool_key)
```
Comment (bquorning) [inline lib/active_record_shards/connection_switcher-4-0.rb]: Potential future optimization:
```rb
connection_handler.connection_pool_list.any? { |pool| pool.spec == connection_pool_key }
```

File: lib/active_record_shards/default_slave_patches.rb [hunk 4]
```diff
@@ -113,20 +110,6 @@ def on_slave_unless_tx
...
-      end
-
-      def exists_with_default_slave?(*args, &block)
-        on_slave_unless_tx { exists_without_default_slave?(*args, &block) }
-      end
-    end
-
     # in rails 4.1+, they create a join class that's used to pull in records for HABTM.
     # this simplifies the hell out of our existence, because all we have to do is inerit on-slave-by-default
     # down from the parent now.
```
Comment (bquorning) [inline lib/active_record_shards/default_slave_patches.rb]: We should rename this module to `HasAndBelongsToManyBuilderExtension`.

**Comment by bquorning** [comment]
Merged and released. Bug reports are welcome.

**Pull request merged**